bin/node/cli/src/service.rs

```diff
@@ -174,6 +174,26 @@ macro_rules! new_full {
 
 			let babe = sc_consensus_babe::start_babe(babe_config)?;
 			service.spawn_essential_task("babe-proposer", babe);
+		}
+
+		// Spawn authority discovery module.
+		if matches!(role, sc_service::config::Role::Authority{..} | sc_service::config::Role::Sentry{..}) {
+			let sentries;
+			let authority_discovery_role;
+
+			match role {
+				sc_service::config::Role::Authority { ref sentry_nodes } => {
+					sentries = sentry_nodes.clone();
+					authority_discovery_role = sc_authority_discovery::Role::Authority (
+						service.keystore(),
+					);
+				}
+				sc_service::config::Role::Sentry {..} => {
+					sentries = vec![];
+					authority_discovery_role = sc_authority_discovery::Role::Sentry;
+				}
+				_ => unreachable!("Due to outer matches! constraint; qed.")
+			}
```

Comment:
I did have `AuthorityDiscovery::new` not take `sc_service::config::Role` for two reasons:

- Make the authority discovery module independent of the underlying network.

- Make it impossible at compile time to pass a key store as a sentry or to not pass a key store as a validator.

Does the above sound reasonable @andresilva ? If you think the additional code in `cli/service` is not worth the benefits above I can move it into the authority discovery module.

Reply:
Yeah, I didn't think of the `sc_service` dependency and also appreciate that keystore requirement is validated statically. Let's keep it in `cli/service` then.
I still prefer the code as written in my suggestion (I didn't test it, just wrote it in Github), mainly because I prefer expressions over statements. :blush:

Reply:
Ah, sorry I that I oversaw your suggestion @andresilva. Indeed a lot better. https://github.com/paritytech/substrate/pull/5568/commits/57d141d6ae2fe170c2962fa46cc5a8d2c53efb50.